### DISCUSSION TITLE
Are there default quota settings in a Strimzi Kafka cluster?

### DISCUSSION THREAD

**Discussion by jslusher** [strimzi/strimzi-kafka-operator#10227] [Q&A]
Strimzi version: 0.36.0
Kafka version: 3.4.0

I'm noticing there are a ton of these logs generated in our Kafka cluster: 

```
INFO Removing CONTROLLER_MUTATION quota for user my-one-user (kafka.server.ControllerMutationQuotaManager) [/config/changes-event-process-thread]"
INFO Sensor for user my-one-user already exists. Changing quota to 9.223372036854776E18 in MetricConfig (kafka.server.ClientRequestQuotaManager) [/config/changes-event-process-thread]"
INFO Removing REQUEST quota for user my-one-user (kafka.server.ClientRequestQuotaManager) [/config/changes-event-process-thread]"
INFO Sensor for user my-one-user already exists. Changing quota to 9.223372036854776E18 in MetricConfig (kafka.server.ClientQuotaManager) [/config/changes-event-process-thread]"
INFO Removing FETCH quota for user my-one-user (kafka.server.ClientQuotaManager) [/config/changes-event-process-thread]"
INFO Sensor for user my-one-user already exists. Changing quota to 9.223372036854776E18 in MetricConfig (kafka.server.ClientQuotaManager) [/config/changes-event-process-thread]"
INFO Removing PRODUCE quota for user my-one-user (kafka.server.ClientQuotaManager) [/config/changes-event-process-thread]"
INFO Removing CONTROLLER_MUTATION quota for user my-other-user (kafka.server.ControllerMutationQuotaManager) [/config/changes-event-process-thread]"
....
```

Are there default quotas set for users? The documentation that I can find about it seems like there shouldn't be, but I can't tell for certain. We're using the KafkaUser Strimzi rsource for what it's worth.

**Comment by scholzj** [comment]
Support for default Kafka quotas is in progress in #9895. Up to Strimzi 0.38, Strimzi would have deleted the default quotas if you set them. From 0.38 on (#9099), it would ignore them if you set them manually through some API.

**Reply by jslusher** [reply]
So I guess my first question is, are the logs above an indication that there are in fact quotas set on our cluster? 

We haven't set quotas of any kind on this cluster to my knowledge. From what I can see in the cluster configuration we don't have anything explicitly set. Most of the users on the cluster are part of a Kafka Connect cluster. Is it possible that Kafka Connect is setting them? Are there quotas that come shipped with the cluster by default?

**Reply by scholzj** [reply]
To be honest, the log snippet is missing a context, so it is hard to know for sure. But the number mentioned there looks like Long.MAX_VALUE. So I read the log as that it is setting the metrics about quotas for this user to the max long value to indicate there is no quota?

**Reply by jslusher** [reply]
I can't say for sure. Looking for context around these logs and it's hard for me to say what's relevant. I can say there are thousands of lines of logs of this type that get logged every few minutes or less. I attached some logs with a little more context in case you might be able to tell me why there are so many of these logs on the server. It's strange to me that there's all this mention of quotas if there aren't any quotas actually set. 
[user-quota-logging-scrubbed.csv](https://github.com/user-attachments/files/15844665/user-quota-logging-scrubbed.csv)

**Reply by scholzj** [reply]
I gave it a quick try - but with Strimzi 0.41 / Kafka 3.7.0 I do not see these messages. So it would be related either to the version or to the settings you have somewhere.

**Reply by jslusher** [reply]
I ran `bin/kafka-configs.sh --describe --entity-type users` and I didn't see any quota settings in the output. Based on the Confluent [documentation about setting quotas](https://docs.confluent.io/platform/current/kafka/post-deployment.html#decrease-the-replication-factor) I should see quota configuration in the output, so I'm going to take that to mean that the logs in my clusters are not caused by quota settings. I'm not sure what other settings to check. I'm starting to wonder if this is behavior caused by clients of a Kafka Connect cluster since we have a lot of them and they all have several connectors. Anyway, thanks for checking!

**Reply by scholzj** [reply]
Well, if you have the User Operator enabled, that would delete any quotas someone would set directly. So you would need to run the command exactly in the time between something sets the quotas and the User oPerator deletes them. I'm pretty sure the User Operator would log the quotas deletion in the DEBUG level logs. But as I said, I can imagine this being just a harmless periodic update to the quota metrics or something. I just find it weird that it is not visible in newer versions -> so either it changed in Kafka or yu have some custom logging configuration.

**Reply by jslusher** [reply]
What's suspicious to me is the logs that say `Removing X quota for user ...`. I hadn't inspected the User operator to see if there was anything in there that might give an indication. I see a lot of log entries like this in the user operator:
```
2024-06-18 17:45:59 WARN  StatusUtils:151 - Reconciliation #379718(timer) KafkaUser(strimzi-kafka-cluster/REDACTED): Contains object at path spec.authorization.acls.resource with unknown properties: name, patternType
```

I tried using `watch` to run several `bin/kafka-configs.sh --describe --entity-type users` commands in a row but I haven't seen any changes to the output so far.

I do see a correlation between reconciliations in the User operator and the logs mentioning quotas in the Kafka cluster, so I do think the User operator is causing all the logging about quota, I just haven't been able to tell if it's actually removing a quota that got set somewhere else or if the logs are standard for some other reason.

**Reply by scholzj** [reply]
> I see a lot of log entries like this in the user operator:

That suggests you have some KafkaUser resources with some invalid YAML I guess.

**Reply by jslusher** [reply]
Those logs are most likely because I set `name` and `patternType` in my `type: cluster` ACL resources:

```
      resource:
        name: kafka-cluster
        patternType: literal
        type: cluster
```
I'm going to clean those up for what it's worth. I'm skeptical that it will help with the quota logging, but we'll see.

**Reply by jslusher** [reply]
Confirmed that removing `patternType` and `name` from user ACLs got rid of those WARN messages. Still plenty of the quota messages though :\

**Reply by scholzj** [reply]
Well, that warning message definitely didn't had any relation to setting quotas, so no, that would not change anything.
